src/censored.jl

```diff
@@ -0,0 +1,343 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The density function (or mass function for discrete `d`) of the censored distribution is
+```math
+f(x; d, l, u) = \\begin{cases}
+    P_d(x \\le l), & x \\le l \\\\
+    P_d(x),         & l < x < u \\\\
+    P_d(x \\ge u), & x \\ge u \\\\
+  \\end{cases},
+```
+where ``P_d(x)`` is the density (or mass) function of `d`, and ``P_d(x \\le b)`` is the
+cumulative distribution function of ``d`` evaluated at ``x = b``.
+If ``x`` is a random variable from ``d``, then `clamp(x, l, u)` is a random variable from
+its censored version. Note that this implies that even if ``d`` is continuous, its censored
+form has discrete mixture components at the bounds.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Union{Missing,Real}, u::Union{Missing,Real})
+    return Censored(d, l, u)
+end
```

Comment:
Maybe we should still handle promotions here instead of in `Censored` only?

```suggestion
`censored(d::D, l::T, u::T) where {T <: Real}`, `censored(d::D, ::Missing, u::Real)`, and `censored(d::D, l::Real, ::Missing)`, or a subset thereof, should be implemented.
"""
function censored(d::UnivariateDistribution, l::T, u::T) where {T<:Real}
    return Censored(d, l, u)
end
function censored(d::UnivariateDistribution, ::Missing, u::Real)
    return Censored(d, missing, u)
end
function censored(d::UnivariateDistribution, l::Real, ::Missing)
    return Censored(d, l, missing)
end

censored(d::UnivariateDistribution, l::Real, u::Real) = censored(d, promote(l, u)...)
censored(d::UnivariateDistribution, ::Missing, ::Missing) = d
```

Reply:
I was trying to avoid the case where a user would need to overload 3 signatures if they had a custom distribution that work also work as a left-, right-, and interval-censored distribution. But that's probably pretty unlikely, so this is better.